lib/manager/pub/readme.md

```diff
@@ -0,0 +1,277 @@
+  ```yaml
+  dependencies:
+    transmogrify:
+      hosted:
+        name: transmogrify
+        url: http://your-package-server.com
+      version: ^1.4.0
+  ```
+
+---
+
+#### Describe which types of dependencies above are supported and which will be implemented in future:
+
+All
+
+## Versioning
+
+#### What versioning scheme do the package files use?
+
+<https://www.dartlang.org/tools/pub/dependencies>
+
+---
+
+#### Does this versioning scheme support range constraints, e.g. `^1.0.0` or `1.x`?
+
+Yes
+
+---
+
+#### Is this package manager used for applications, libraries, or both? If both, is there a way to tell which is which?
+
+Both. No.
+
+---
+
+#### If ranges are supported, are there any cases when Renovate should pin ranges to exact versions if rangeStrategy=auto?
+
+Dev dependencies can always be pinned.
+
+## Lookup
+
+#### Is a new datasource required? Provide details
+
+Not sure.
+
+---
+
+#### Will users need the capability to specify a custom host/registry to look up? Can it be found within the package files, or within other files inside the repository, or would it require Renovate configuration?
+
+Yes
+
+A registry can be specified in _pubspec.yaml_ with:
+
+```yaml
+dependencies:
+  transmogrify:
+    hosted:
+      name: transmogrify
+      url: http://your-package-server.com
+    version: ^1.4.0
+```
+
+---
+
+#### Do the package files contain any "constraints" on the parent language (e.g. supports only v3.x of Python) or platform (Linux, Windows, etc) that should be used in the lookup procedure?
+
+Yes, SDK version can be defined using:
+
+```yaml
+environment:
+  sdk: '>=2.0.0 <3.0.0'
+```
+
+---
+
+#### Will users need the ability to configure language or other constraints using Renovate config?
+
+No
+
+## Artifacts
+
+#### Are lock files or checksum files used? Mandatory?
+
+Yes and yes, in _pubspec.lock_.
+
+---
+
+#### If so, what tool and exact commands should be used if updating 1 or more package versions in a dependency file?
+
+Update all dependencies:
+
+`flutter packages upgrade`
+
+Update 1 dependency is not supported by a command. The package needs to be changed with the version number in _pubspec.yaml_ and then run `flutter packages upgrade`.
+
+---
+
+#### If applicable, describe how the tool maintains a cache and if it can be controlled via CLI or env? Do you recommend the cache be kept or disabled/ignored?
+
+- `PUB_CACHE` environment variable defines the cache's location: <https://www.dartlang.org/tools/pub/environment-variables>
+
+---
+
+#### If applicable, what command should be used to generate a lock file from scratch if you already have a package file? This will be used for "lock file maintenance".
+
+`flutter packages get`
+
+## Other
+
+#### Is there anything else to know about this package manager?
+
+The information here refers to the Flutter SDK package manager, which is built on top of Dart's `pub` package manager but differs in some ways.
```

Comment:
I hadn't realised there were two approaches here. Can you elaborate on the impact of this? e.g. should we support pub and flutter separately within Renovate, even if they're mostly the same?

Reply:
My understanding is that Flutter's package manager is a superset of Dart's `pub`. 

It adds the Flutter SDK to the spec:
```yaml
dependencies:
  flutter_driver:
    sdk: flutter
    version: ^0.0.1
```

Also, the commands for upgrading packages are different, while in Dart it's `pub ...` in Flutter it's `flutter ...`.

I think it's probably better if they are separate because of these differences.

Reply:
`flutter packages get` `flutter packages upgrade` is equivalent to
```
export FLUTTER_ROOT="/Users/my_user/flutter"
pub get # or pub upgrade
```

Reply:
As far as I know support for `sdk: flutter` is built in to `pub`, and is enabled by setting `FLUTTER_ROOT` env variable.